content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+educational use cases like ours since it allows anyone with a web
+browser to quickly try Python and its huge ecosystem of packages without
+any installation or system maintenance.
+
+
+## How did we port it and what issues did we face?
+
+It was remarkably pleasant to port Pandas Tutor over to Pyodide from our
+original server-based setup, even as first-time Pyodide users. We barely
+had to change our code at all.
+
+
+### Step 1: create a self-contained Pandas Tutor wheel
+
+The recommended way to load custom Python code into Pyodide is [by using
+a
+wheel](https://pyodide.org/en/stable/usage/loading-custom-python-code.html),
+so we first created a wheel to package up the Pandas Tutor backend
+code. That was straightforward since it's pure-Python and doesn't
+require any WASM code compilation.
+
+However, we ran into our first issue right away: Since Pandas Tutor has
+almost a dozen package dependencies, we originally thought we could use
+`micropip` to install them all, as the [Pyodide documentation
+recommends](https://pyodide.org/en/stable/usage/loading-packages.html).
+Unfortunately, some packages couldn't be installed with `micropip` since
+they didn't have wheels available on PyPI (e.g.,
+`micropip.install('docopt')` doesn't work). Our solution was to bundle
+(i.e., inline) all dependencies into a self-contained Pandas
+Tutor wheel so that our web frontend code needs only one call to
+`micropip` instead of almost a dozen separate calls:
+
+```js
+await pyodide.runPythonAsync(`
+import micropip
+await micropip.install('https://pandastutor.com/build/pandastutor-1.0-py3-none-any.whl')
+`);
+```
+
+This approach also has the nice side-effect of pinning all the
+dependency versions for better API stability. The only dependency we
+didn't bundle was pandas because that requires WASM-compiled code. But
+fortunately Pyodide comes with pandas
+[built-in](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)!
+
+
+### Step 2: create a new Pyodide entry point
+
+The only modification we had to make to the Pandas Tutor codebase was
+adding a new `pyodide_main.py` file as the entry point for Pyodide:
+
+```py
+# simplified contents of pandas_tutor/pyodide_main.py
+from .main import run_user_code
+import js # to interface with pyodide
+def run_code_from_js():
+    return run_user_code(js.globalUserCode)
+```
+
+This entry point uses the `js` module to access JavaScript global
+variables. On the JavaScript end of our web frontend code, we created a
+corresponding function that takes the user's code to run (`userCode`)
+and calls the `run_code_from_js()` Python function:
+
+```js
+function runPyodidePandastutor(userCode) {
+  window.globalUserCode = userCode; // pass function argument as a global variable
+  const result = pyodide.runPython('pandas_tutor.pyodide_main.run_code_from_js()');
+  window.globalUserCode = undefined; // null this out afterward to prevent leakage
+  return result;
+}
```

Comment:
Normally, something like the following would be the way to do it,

```suggestion
function runPyodidePandastutor(userCode) {
  const pandas_tutor_py = pyodide.pyimport("pandas_tutor");
  // assuming main.py is imported in `pandas_tutor.__init__.py` or directly
  // pyimport("pandas_tutor.main")
  const result = pandas_tutor_py.main.run_user_code(user_code);  
  return result;
}
```
and maybe moving the pyimport call outside of the function to avoid running it for each call.

Yeah, ideally it would be nice to make it a bit more elegant since we are showing it in the blog post. If the above doesn't work, let me know and we will find another solution.